control-plane/agents/src/bin/core/volume/operations.rs

```diff
@@ -738,3 +665,282 @@ impl ResourceShutdownOperations for OperationGuardArc<VolumeSpec> {
+                    Ok(mut replica) => {
+                        let request = DestroyReplica::from(replica_state.clone());
+                        replica.destroy(registry, &request.with_disown_all()).await
+                    }
+                    Err(error) => Err(error),
+                };
+                if let Err(error) = result {
+                    volume_clone.error(&format!(
+                        "Failed to delete replica {:?} from volume, error: {}",
+                        replica_state,
+                        error.full_string()
+                    ));
+                }
+            }
+            Err(SvcError::ReplicaCreateNumber {
+                id: request.uuid.to_string(),
+            })
+        } else {
+            Ok(())
+        };
+
+        // we can destroy volume on error because there's no volume resource created on the nodes,
+        // only sub-resources (such as nexuses/replicas which will be garbage-collected later).
+        volume
+            .complete_create(result, registry, OnCreateFail::Delete)
+            .await?;
+        Ok(volume)
+    }
+}
+
+#[async_trait::async_trait]
+impl ResourceLifecycleExt<CreateVolumeSource<'_>> for OperationGuardArc<VolumeSpec> {
+    type CreateOutput = Self;
+
+    async fn create_ext(
+        registry: &Registry,
+        request_src: &CreateVolumeSource,
+    ) -> Result<Self::CreateOutput, SvcError> {
+        let request = request_src.source();
+        match request_src {
+            CreateVolumeSource::Pool(params) => params.pre_flight_check()?,
+            CreateVolumeSource::Snapshot(params) => params.pre_flight_check()?,
+        }
+
+        let specs = registry.specs();
+        let mut volume = specs
+            .get_or_create_volume(request)
+            .operation_guard_wait()
+            .await?;
+        let volume_clone = volume.start_create_update(registry, request).await?;
+
+        // If the volume is a part of the ag, create or update accordingly.
+        registry.specs().get_or_create_affinity_group(&volume_clone);
+
+        let context = Context {
+            registry,
+            volume: &mut volume,
+        };
+        let result = match request_src {
+            CreateVolumeSource::Pool(params) => params.run(context).await,
+            CreateVolumeSource::Snapshot(params) => params.run(context).await,
+        };
+
+        // we can destroy volume on error because there's no volume resource created on the nodes,
+        // only sub-resources (such as nexuses/replicas which will be garbage-collected later).
+        volume
+            .complete_create(result, registry, OnCreateFail::Delete)
+            .await?;
+        Ok(volume)
+    }
+}
+
+/// A volume can be created with different sources for its replicas.
+pub(super) enum CreateVolumeSource<'a> {
+    /// Carve out new replicas from a pool matching the requested topology.
+    Pool(&'a CreateVolume),
```

Comment:
Shouldn't his be `None` rather than `Pool` ?

Reply:
I wasn't sure what to call it, I'll rename it :+1: